woocommerce/woocommerce-ios
# [Order Creation] Search Products by SKU

Closes: #


## Description

With this PR we add support to search products by SKU when creating an order, something that has been requested by our users a few times.
There have been a few challenges when adapting the SKU search to this case. First and foremost, the products endpoint returns Product Variations as products when returning its data. Because of this, we have to convert it to `ProductVariation` if required, so we can handle them as we were doing before in the `ProductSelectorViewModel`. Furthermore, we show the product variations together with products in the same list as we do in the Products tab, unlike all the other cases in the Product Selector that have a separate screen for all the product variations. This brings more easiness to the user when selecting by SKU, as they can search also variations and select them right away, without having to open a second screen.

Apart from that, we change the logic when the item is deleted from the order detail screen. Before it relied on the product selector to remove it and call back. This raised a bug when the product selector didn't have a reference to the product or variation (e.g. the product page wasn't loaded yet so it wasn't in the database). Now we delete it automatically from the order detail screen. (See changes in the `EditableOrderViewModel`)

Moreover, I took this opportunity to refactor `ProductSelectorViewModel` to collapse all the ID array properties we had into one, encapsulating product and variations alike. Apart from that, I tweaked the sync logic to show the loading screen when we trigger a new search, making it more visible to the users.

## Testing instructions


As the logic for Product Selection changed a bit, please perform a regression test for all the Product Selection cases including editing the order from the order details screen -deleting and changing the quantity of items-. Please also test all the following cases:

- Search SKU for a product and add it. See that the confirm button is updated. Go back to the default state -with no search- and make sure that the product is selected. 
- Search SKU for a variation and add it. See that the confirm button is updated. Go back to the default state -with no search- and make sure that the variation is selected. 
- Deselect products and variations while in SKU search and default mode.

See also that the search is tracked with the right filter:

`🔵 Tracked order_creation_product_selector_search_triggered, properties: [AnyHashable("blog_id"): 205293073, AnyHashable("is_wpcom_store"): true, AnyHashable("search_filter"): "all"]`

and

`🔵 Tracked order_creation_product_selector_search_triggered, properties: [AnyHashable("blog_id"): 205293073, AnyHashable("is_wpcom_store"): true, AnyHashable("search_filter"): "sku"]`

## Screenshots


https://github.com/woocommerce/woocommerce-ios/assets/1864060/0672c32e-7a15-4873-898f-8c3471d1bced



---
